[quant][docs] Add more docs for quantization.rst

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #76223
* __->__ #75998

Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35739111](https://our.internmc.facebook.com/intern/diff/D35739111)

jerryzh168 committed:
[quant][docs] Add more docs for quantization.rst

Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35739111).

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -38,20 +38,177 @@ that perform all or part of the computation in lower precision. Higher-level
 to lower precision with minimal accuracy loss.
 
-Quantization requires users to be aware of three concepts:
+Quantization Stack
```
would it make sense to put the information for the common user (don't care about internals, just want to know how to quantize something with default settings) at the top, and put the details at the bottom?  Seems like this is currently doing the reverse.

vkuzo commented on docs/source/quantization.rst:
```diff
@@ -38,20 +38,177 @@ that perform all or part of the computation in lower precision. Higher-level
+~~~~~~~~~~~~~~~~~
+We need a representation for quantized activations and weights in a quantized model. A Quantized Tensor is a type of a PyTorch Tensor. It is quantized from a Tensor with higher precisions like float. And it stores quantization parameters like scale and zero_point.
 
-Quantization API Summary
```
IMO the API summary should be as close to the top as possible, and people shoudl scroll down if they need more details

vkuzo reviewed (approved):
accepting to unblock, it would be nice to keep the API summary as close to the top of the page as possible

dzdang commented on docs/source/quantization.rst:
```diff
@@ -14,12 +14,12 @@ Introduction to Quantization
 tensors at lower bitwidths than floating point precision. A quantized model
-executes some or all of the operations on tensors with integers rather than
+executes some or all of the operations on tensors with reduced precision rather than
 floating point values. This allows for a more compact model representation and
```
"rather than floating point values" should this be changed, since you changed the wording from integers->reduced precision. I thought we have some fp16 quantization stuff?

jerryzh168 replied:
makes sense, I think floating point should be full precision in this case

jerryzh168 replied:
yeah fp16 is also considered reduced precision

dzdang commented on docs/source/quantization.rst:
```diff
@@ -38,20 +38,177 @@ that perform all or part of the computation in lower precision. Higher-level
 
-Quantization API Summary
----------------------------------------
+PyTorch supports both per tensor and per channel symmetric and asymmetric quantization. Per tensor means that all the values within the tensor are quantized the same way with the same quantization parameters. Per channel means that for each dimension, typically the channel dimension of a tensor, the values in the tensor are quantized with difference quantization parameters. This allows for less e ...
```
different*

"This allows for less error in converting tensors to quantized values." Not too familiar with this, but is there justification we can provide for this statement? If not, I think we may need to be less definitively stated, e.g., "This flexibility may allow for less error in converting tensors to quantized values as scaling and zero_point parameters may be tailored for each slice of the desired dimension"

jerryzh168 replied:
yes, I can add some explanations, basically per channel is more accurate because when each channel of values is quantized with their own quantization parameters, if there is some outlier, the effect would be restricted to that one channel, instead of impacting the entire tensor

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35739111](https://our.internmc.facebook.com/intern/diff/D35739111)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35739111](https://our.internmc.facebook.com/intern/diff/D35739111)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35739111](https://our.internmc.facebook.com/intern/diff/D35739111)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35739111](https://our.internmc.facebook.com/intern/diff/D35739111)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35739111](https://our.internmc.facebook.com/intern/diff/D35739111)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][docs] Add more docs for quantization.rst"


Summary:
Add more details to user facing docs quantization.rst, which will be displayed in the official quantization doc page: https://pytorch.org/docs/stable/quantization.html
This includes:
* docs for quantization stack (quantized tensor, quantized operator and modules, observer, fake_quantize, QConfig, quantization flow)
* Added support table for quantization mode, quantization flow mode and backend, (also moved around operator support table)
* restructured eager mode and fx mode docs as well

Test Plan:
inspect the doc that's built by github ci

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D35739111](https://our.internmc.facebook.com/intern/diff/D35739111)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35739111).
